### PR TITLE
[optim] lbfgs: handle complex params as independent real params

### LINKED ISSUE #118148: Enable complex dtype support for LBFGS
### 🚀 The feature, motivation and pitch

Today, there is no complex support for LBFGS.  Adding it should be doable with `view_as_real` similar to how we implement complex support for other optimizers.

### Alternatives

_No response_

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @crcrpar

### PR DESCRIPTION
**Pull request by fzimmermann89**
Ref: #86340

Fixes #118148 

This fixes LBFGS for complex parameters. Complex parameters are handled as R^2.
I also added a test, unfortunately, due to the closure required, I could not use the existing `_test_complex_optimizer` used for all other optimizers.
Lbfgs is special, as it will call the objective function multiple times internally. So I felt making a one-off test for lbfgs might be justifiable. 
We will test if each step taken internally by the optimizer is the same for R^2 and complex parameters.

Let me know if the approach is ok, thanks

### CODE REVIEW

File: test/optim/test_optim.py [hunk 1]
```diff
@@ -834,6 +835,49 @@ def test_rprop(self):
...
+            lambda weight, bias: LBFGS(
+                [weight, bias], line_search_fn="strong_wolfe"
+            ),
+            ignore_multidevice=True,
+        )
+
+
+    def test_lbfgs_complex(self):
+        for linesearch in ("strong_wolfe", None):
+            complex_params = [torch.randn(10, 5, dtype=torch.complex64, requires_grad=True), torch.randn(10, dtype=torch.complex64, requires_grad=True), torch.randn(10, 5, dtype=torch.float32, requires_grad=True)]
```
Comment (janeyx99) [inline test/optim/test_optim.py:852]: is the last param being float32 intentional?

**Review by janeyx99 — COMMENTED** [review]
Thank you for taking this on! Looks pretty logical overall :D Left some comments--there is some churn in the test files as I'm migrating the complex tests at this very moment.

File: test/optim/test_optim.py (hunk 1 above)
Comment (fzimmermann89) [reply test/optim/test_optim.py:852]: Yes. 
We would like to use bfgs with both complex and real parameters. Our first try to fix the issue did not take this into account and it failed when real and complex gradients were concatenated in _gather_flat_grad.

File: test/optim/test_optim.py [hunk 2]
```diff
@@ -834,6 +835,49 @@ def test_rprop(self):
...
+        )
+        self._test_basic_cases(
+            lambda weight, bias: LBFGS(
+                [weight, bias], line_search_fn="strong_wolfe"
+            ),
+            ignore_multidevice=True,
+        )
+
+
+    def test_lbfgs_complex(self):
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply test/optim/test_optim.py:837]: - IMHO it is cleaner to leave it as a separate test than to use conditionals inside test_complex,  the body of the test is different -- they only share some setup lines. So t save a logic duplication of 5 lines, we would have a conditional switching between two different implementations of the inner loop. 
- I tried to match the style of the rest of the file. I could have used` @parametrize`, but this would have been different than in all other test in this file. Not sure what you meant by `@optims` as this would only parameterize for different optimizers, right? 
So I could use `@parametrize / instantiate_parametrized_tests` instead of the loop, or match the current style, or move the existing tests over to parameterize as well.
Not sure what your plans are for the remaining tests, i.e. which ones you  plan to move into the `TestOptimRenewed`

File: test/optim/test_optim.py [hunk 3]
```diff
@@ -834,6 +835,49 @@ def test_rprop(self):
...
+                    p.grad = torch.randn_like(p)
+                    steps.append(p.detach().clone())
+                    grads_losses.append(p.grad.clone())
+                grads_losses.append(torch.randn(1))
+                return grads_losses[-1].clone()
+
+            def complex_closure():
+                for p in complex_params:
+                    if torch.is_complex(p):
+                        self.assertEqual(torch.view_as_real(p), steps.pop(0))
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply test/optim/test_optim.py]: moved it to the end of the test, outside the closures.

File: test/optim/test_optim.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/optim/test_optim.py:837]: `@optims` would allow you to get all the configs for LBFGS so you wouldn't have to specify the forloop toggling linesearch. This way, if new features get added to LBFGS, this test would run automatically for those new configs without changing the test. This would also allow ease of testing across different devices and dtypes. You would want something like

```
@optims([optim for optim in optim_db if optim.optim_cls.__name__ == "LBFGS"], dtypes=[torch.float32])
def test_lbfgs_complex(self, device, dtype)
```

I'm okay with this being a separate test (though I like the idea of mixing in a non-complex param for the other test_complex + would be a change for later). If we want to do that, we should change the skip message in common_optimizers for test_complex for LBFGS to say that it's tested separately through this test.

File: test/optim/test_optim.py [hunk 4]
```diff
@@ -834,6 +834,40 @@ def test_rprop(self):
             )
 
 
+    def test_lbfgs_complex(self):
+        for linesearch in ("strong_wolfe", None):
+            complex_params = [torch.randn(10, 5, dtype=torch.complex64, requires_grad=True), torch.randn(10, dtype=torch.complex64, requires_grad=True), torch.randn(10, 5, dtype=torch.float32, requires_grad=True)]
+            real_params = [torch.view_as_real(param).detach().clone().requires_grad_() if param.is_complex() else param.detach().clone().requires_grad_() for param in complex_params]
+            complex_steps = []
+            real_steps = []
```
Comment (janeyx99) [inline test/optim/test_optim.py:842]:
```suggestion
```

File: test/optim/test_optim.py [hunk 5]
```diff
@@ -834,6 +834,40 @@ def test_rprop(self):
...
+            grads_losses = []
+
+            def real_closure():
+                for p in real_params:
+                    p.grad = torch.randn_like(p)
+                    grads_losses.append(p.grad.clone())
+                    real_steps.append(p.clone())
```
Comment (janeyx99) [inline test/optim/test_optim.py:849]:
```suggestion
```

File: test/optim/test_optim.py [hunk 6]
```diff
@@ -834,6 +834,40 @@ def test_rprop(self):
...
+                grads_losses.append(torch.randn(1))
+                return grads_losses[-1].clone()
+
+            def complex_closure():
+                for p in complex_params:
+                    if torch.is_complex(p):
+                        p.grad = torch.view_as_complex(grads_losses.pop(0))
+                        complex_steps.append(torch.view_as_real_copy(p))
```
Comment (janeyx99) [inline test/optim/test_optim.py:857]:
```suggestion
```

File: test/optim/test_optim.py [hunk 7]
```diff
@@ -834,6 +834,40 @@ def test_rprop(self):
...
+                    else:
+                        p.grad = grads_losses.pop(0)
+                        complex_steps.append(p.clone())
```
Comment (janeyx99) [inline test/optim/test_optim.py:860]:
```suggestion
```

File: test/optim/test_optim.py [hunk 8]
```diff
@@ -834,6 +834,40 @@ def test_rprop(self):
...
+                return grads_losses.pop(0)
+
+            real_optim = LBFGS(real_params, line_search_fn=linesearch)
+            complex_optim = LBFGS(complex_params, line_search_fn=linesearch)
+            real_optim.step(real_closure)
+            complex_optim.step(complex_closure)
+            self.assertEqual(complex_steps, real_steps)
```
Comment (janeyx99) [inline test/optim/test_optim.py:867]:
```suggestion
            self.assertEqual(complex_params, real_params)
```
Why not just this?

Also, could we step multiple times? Like 3?

**Comment by janeyx99** [comment]
Also, I think you will need to rebase to the latest changes

**Review by janeyx99 — COMMENTED** [review]
My plans are to move all tests from TestOptim to TestOptimRenewed in the next few months.

**Comment by fzimmermann89** [comment]
The idea was that all intermediate steps in the line-search should also be the same, as it it could be that both for real and it could be that we end up at the same result using random, but different paths, which would break the specification.

But if we don't care about that, I can also make the normal TestOptimRenewed.test_complex work for LBFGS.

It seems that there is a preference for conditionals on the cls name, for both SparseAdam and LBFGS, instead of using
OptimInfo.only_supports_sparse_grads/step_requires_closure.
Do you think this is intentional or will it be refactored? 
Before I again commit something matching the current style of main before a refactor..

**Comment by janeyx99** [comment]
> The idea was that all intermediate steps in the line-search should also be the same, as it it could be that both for real and it could be that we end up at the same result using random, but different paths, which would break the specification.

Ah, I see, this sounds better to me—could you add a comment for posterity then?

> It seems that there is a preference for conditionals on the cls name, for both SparseAdam and LBFGS, instead of using OptimInfo.only_supports_sparse_grads/step_requires_closure. Do you think this is intentional or will it be refactored? Before I again commit something matching the current style of main before a refactor..

this is not intentional, and we should move to using the attributes!

**Review by janeyx99 — APPROVED** [review]
Looks awesome! Thank you :D

File: test/test_optim.py [hunk 9]
```diff
@@ -150,24 +150,59 @@ def test_complex(self, device, dtype, optim_info):
...
+                for p in complex_params:
+                    if torch.is_complex(p):
+                        p.grad = torch.view_as_complex(grads_losses.pop(0))
+                        complex_steps.append(torch.view_as_real_copy(p))
+                    else:
+                        p.grad = grads_losses.pop(0)
+                        complex_steps.append(p.clone())
+                return grads_losses.pop(0)
+
+                for _ in range(3):
```
Comment (janeyx99) [inline test/test_optim.py:191]: This should be indented forward

**Comment by fzimmermann89** [comment]
sorry, I broke stuff, give me a sec ...

**Comment by janeyx99** [comment]
Once CI is all green, feel free to issue the merge comment yourself!
